examples/compiler/class_table.rs

```diff
@@ -1,47 +1,51 @@
 use crate::compiler;
-use salsa::{query_definition, query_prototype};
+use salsa::query_prototype;
 use std::sync::Arc;
 
 query_prototype! {
-    pub trait ClassTableQueryContext: compiler::CompilerQueryContext {
+    pub trait ClassTableDatabase: compiler::CompilerDatabase {
         /// Get the fields.
-        fn fields() for Fields;
+        fn fields(class: DefId) -> Arc<Vec<DefId>> {
+            type Fields;
+        }
 
         /// Get the list of all classes
-        fn all_classes() for AllClasses;
+        fn all_classes(key: ()) -> Arc<Vec<DefId>> {
+            type AllClasses;
+        }
 
         /// Get the list of all fields
-        fn all_fields() for AllFields;
+        fn all_fields(key: ()) -> Arc<Vec<DefId>> {
+            type AllFields;
+        }
     }
 }
 
 #[derive(Copy, Clone, Debug, Hash, PartialEq, Eq)]
 pub struct DefId(usize);
 
-query_definition! {
-    pub AllClasses(_: &impl ClassTableQueryContext, (): ()) -> Arc<Vec<DefId>> {
+impl<DB: ClassTableDatabase> salsa::QueryFunction<DB> for AllClasses {
```

Comment:
Just an idea, but, if we have an impl anyway, then perhaps we can move storage declaration into the impl?

```
trait salsa::QueryFunction<DB> {
    type Storage = MemoizedStorage<Self>;
    fn execute(db: &DB, key: Self::Key) -> Self::Value
}
```

Reply:
Hmm, OK, a few things:

1. I agree it has a boilerplate feel to write the impl. I considered just calling a fn with same name but it has a kind of "magic" feeling to me -- there is no "explicit connection". OTOH, convenient.  I also liked the idea that you can put the impl wherever it seems to fit best (it doesn't have to be in the same module).
2. I'd rather not ask the user to pick the storage via an associated type because I think the type is a bit complex to type.
3. Also, you don't need the impl for the "input" case.

Regarding jump-to-def, it seems like if you are on `db.my_query(foo)` and you hit jump-to-def, a ctags like system might take you to `fn my_query`, but a more precise system would take you to the macro (which defines the trait).

If we were going to go with the "calls a fn" option, I would prefer having the option to write `use fn <path>;` that specifies which fn you want (but it defaults to `use fn my_query;`, which has the behavior you described).  I can give that a try. I suspect in the end less boilerplate is better.

Reply:
>Regarding jump-to-def

jump-to-def would work ok, but I am talking about another (subjectively, more useful) feature, when you explicitly type a name of the thing into the input box (as opposed to clicking on the usage), and the editor brings you to the definition. I think that's xref-find-apropos in Emacs. 

> I would prefer having the option to write use fn <path>;

Yes, that was what my `by all_classes;` clause suggested, but `use fn`, as a syntax, seems neater. I think it clears both concerns from 1: explicit connection + ability to put the query impl elsewhere?

Reply:
Oh, I missed the `by all_classes` in your example. I agree it basically clears the concerns.

Reply:
I'll try it.